# <font color='blue'>Week 6 - Machine Learning Intro</font>

### <font color='red'> Used: Python and its libraries; NumPy and Pandas library. Jupyter Notebook. </font>


**XGBoost Algorithm**
XGBoost is one of the most popular machine learning algorithm these days. Regardless of the type of prediction task at hand; regression or classification.

In [1]:
%autosave 20

Autosaving every 20 seconds


Reference links:

https://xgboost.readthedocs.io/en/latest/

https://scikit-learn.org/stable/modules/generated/sklearn.datasets.load_boston.html

https://towardsdatascience.com/linear-regression-on-boston-housing-dataset-f409b7e4a155

https://medium.com/@haydar_ai/learning-data-science-day-9-linear-regression-on-boston-housing-dataset-cd62a80775ef

Import the Boston Housing dataset from scikit-learn and store it in a variable called boston_dataset.

In [3]:
from sklearn.datasets import load_boston
#import sklearn.datasets.load_boston()

In [4]:
boston_dataset = load_boston(return_X_y=False)

In [5]:
boston_dataset

{'data': array([[6.3200e-03, 1.8000e+01, 2.3100e+00, ..., 1.5300e+01, 3.9690e+02,
         4.9800e+00],
        [2.7310e-02, 0.0000e+00, 7.0700e+00, ..., 1.7800e+01, 3.9690e+02,
         9.1400e+00],
        [2.7290e-02, 0.0000e+00, 7.0700e+00, ..., 1.7800e+01, 3.9283e+02,
         4.0300e+00],
        ...,
        [6.0760e-02, 0.0000e+00, 1.1930e+01, ..., 2.1000e+01, 3.9690e+02,
         5.6400e+00],
        [1.0959e-01, 0.0000e+00, 1.1930e+01, ..., 2.1000e+01, 3.9345e+02,
         6.4800e+00],
        [4.7410e-02, 0.0000e+00, 1.1930e+01, ..., 2.1000e+01, 3.9690e+02,
         7.8800e+00]]),
 'target': array([24. , 21.6, 34.7, 33.4, 36.2, 28.7, 22.9, 27.1, 16.5, 18.9, 15. ,
        18.9, 21.7, 20.4, 18.2, 19.9, 23.1, 17.5, 20.2, 18.2, 13.6, 19.6,
        15.2, 14.5, 15.6, 13.9, 16.6, 14.8, 18.4, 21. , 12.7, 14.5, 13.2,
        13.1, 13.5, 18.9, 20. , 21. , 24.7, 30.8, 34.9, 26.6, 25.3, 24.7,
        21.2, 19.3, 20. , 16.6, 14.4, 19.4, 19.7, 20.5, 25. , 23.4, 18.9,
        35.4, 24.7, 3

Print keys of boston dataset

In [6]:
# o/p =dict_keys(['data', 'target', 'feature_names', 'DESCR'])

In [7]:
boston_dataset.keys()

dict_keys(['data', 'target', 'feature_names', 'DESCR', 'filename'])

In [8]:
boston_dataset.data

array([[6.3200e-03, 1.8000e+01, 2.3100e+00, ..., 1.5300e+01, 3.9690e+02,
        4.9800e+00],
       [2.7310e-02, 0.0000e+00, 7.0700e+00, ..., 1.7800e+01, 3.9690e+02,
        9.1400e+00],
       [2.7290e-02, 0.0000e+00, 7.0700e+00, ..., 1.7800e+01, 3.9283e+02,
        4.0300e+00],
       ...,
       [6.0760e-02, 0.0000e+00, 1.1930e+01, ..., 2.1000e+01, 3.9690e+02,
        5.6400e+00],
       [1.0959e-01, 0.0000e+00, 1.1930e+01, ..., 2.1000e+01, 3.9345e+02,
        6.4800e+00],
       [4.7410e-02, 0.0000e+00, 1.1930e+01, ..., 2.1000e+01, 3.9690e+02,
        7.8800e+00]])

In [9]:
boston_dataset.feature_names

array(['CRIM', 'ZN', 'INDUS', 'CHAS', 'NOX', 'RM', 'AGE', 'DIS', 'RAD',
       'TAX', 'PTRATIO', 'B', 'LSTAT'], dtype='<U7')

In [10]:
len(boston_dataset)

5

In [11]:
print(boston_dataset.target)

[24.  21.6 34.7 33.4 36.2 28.7 22.9 27.1 16.5 18.9 15.  18.9 21.7 20.4
 18.2 19.9 23.1 17.5 20.2 18.2 13.6 19.6 15.2 14.5 15.6 13.9 16.6 14.8
 18.4 21.  12.7 14.5 13.2 13.1 13.5 18.9 20.  21.  24.7 30.8 34.9 26.6
 25.3 24.7 21.2 19.3 20.  16.6 14.4 19.4 19.7 20.5 25.  23.4 18.9 35.4
 24.7 31.6 23.3 19.6 18.7 16.  22.2 25.  33.  23.5 19.4 22.  17.4 20.9
 24.2 21.7 22.8 23.4 24.1 21.4 20.  20.8 21.2 20.3 28.  23.9 24.8 22.9
 23.9 26.6 22.5 22.2 23.6 28.7 22.6 22.  22.9 25.  20.6 28.4 21.4 38.7
 43.8 33.2 27.5 26.5 18.6 19.3 20.1 19.5 19.5 20.4 19.8 19.4 21.7 22.8
 18.8 18.7 18.5 18.3 21.2 19.2 20.4 19.3 22.  20.3 20.5 17.3 18.8 21.4
 15.7 16.2 18.  14.3 19.2 19.6 23.  18.4 15.6 18.1 17.4 17.1 13.3 17.8
 14.  14.4 13.4 15.6 11.8 13.8 15.6 14.6 17.8 15.4 21.5 19.6 15.3 19.4
 17.  15.6 13.1 41.3 24.3 23.3 27.  50.  50.  50.  22.7 25.  50.  23.8
 23.8 22.3 17.4 19.1 23.1 23.6 22.6 29.4 23.2 24.6 29.9 37.2 39.8 36.2
 37.9 32.5 26.4 29.6 50.  32.  29.8 34.9 37.  30.5 36.4 31.1 29.1 50.
 33.3 3

In [12]:
print(boston_dataset.DESCR)

.. _boston_dataset:

Boston house prices dataset
---------------------------

**Data Set Characteristics:**  

    :Number of Instances: 506 

    :Number of Attributes: 13 numeric/categorical predictive. Median Value (attribute 14) is usually the target.

    :Attribute Information (in order):
        - CRIM     per capita crime rate by town
        - ZN       proportion of residential land zoned for lots over 25,000 sq.ft.
        - INDUS    proportion of non-retail business acres per town
        - CHAS     Charles River dummy variable (= 1 if tract bounds river; 0 otherwise)
        - NOX      nitric oxides concentration (parts per 10 million)
        - RM       average number of rooms per dwelling
        - AGE      proportion of owner-occupied units built prior to 1940
        - DIS      weighted distances to five Boston employment centres
        - RAD      index of accessibility to radial highways
        - TAX      full-value property-tax rate per $10,000
        - PTRATIO  pu

# This is not a data frame!
## so we ned to create a df DataFrame NOW from DataSet.

In [13]:
import pandas

In [14]:
# this is wrong, it will have no data as .data is missing
pandas.DataFrame(data=boston_dataset, columns=boston_dataset.feature_names)
# this is wrong, it will have no data as .data is missing

Empty DataFrame
Columns: [CRIM, ZN, INDUS, CHAS, NOX, RM, AGE, DIS, RAD, TAX, PTRATIO, B, LSTAT]
Index: []

In [15]:
df_boston = pandas.DataFrame(data=boston_dataset.data, columns=boston_dataset.feature_names)
df_boston

CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD    TAX  \
0    0.00632  18.0   2.31   0.0  0.538  6.575  65.2  4.0900  1.0  296.0   
1    0.02731   0.0   7.07   0.0  0.469  6.421  78.9  4.9671  2.0  242.0   
2    0.02729   0.0   7.07   0.0  0.469  7.185  61.1  4.9671  2.0  242.0   
3    0.03237   0.0   2.18   0.0  0.458  6.998  45.8  6.0622  3.0  222.0   
4    0.06905   0.0   2.18   0.0  0.458  7.147  54.2  6.0622  3.0  222.0   
..       ...   ...    ...   ...    ...    ...   ...     ...  ...    ...   
501  0.06263   0.0  11.93   0.0  0.573  6.593  69.1  2.4786  1.0  273.0   
502  0.04527   0.0  11.93   0.0  0.573  6.120  76.7  2.2875  1.0  273.0   
503  0.06076   0.0  11.93   0.0  0.573  6.976  91.0  2.1675  1.0  273.0   
504  0.10959   0.0  11.93   0.0  0.573  6.794  89.3  2.3889  1.0  273.0   
505  0.04741   0.0  11.93   0.0  0.573  6.030  80.8  2.5050  1.0  273.0   

     PTRATIO       B  LSTAT  
0       15.3  396.90   4.98  
1       17.8  396.90   9.14  
2       17.8  392.83   4.03  
3       18.7  394.63   2.94  
4       18.7  396.90   5.33  
..       ...     ...    ...  
501     21.0  391.99   9.67  
502     21.0  396.90   9.08  
503     21.0  396.90   5.64  
504     21.0  393.45   6.48  
505     21.0  396.90   7.88  

[506 rows x 13 columns]

Print dimensions of this dataset

In [16]:
# o/p =(506, 13)

In [17]:
df_boston.shape

(506, 13)

In [18]:
df_boston.describe

<bound method NDFrame.describe of         CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD    TAX  \
0    0.00632  18.0   2.31   0.0  0.538  6.575  65.2  4.0900  1.0  296.0   
1    0.02731   0.0   7.07   0.0  0.469  6.421  78.9  4.9671  2.0  242.0   
2    0.02729   0.0   7.07   0.0  0.469  7.185  61.1  4.9671  2.0  242.0   
3    0.03237   0.0   2.18   0.0  0.458  6.998  45.8  6.0622  3.0  222.0   
4    0.06905   0.0   2.18   0.0  0.458  7.147  54.2  6.0622  3.0  222.0   
..       ...   ...    ...   ...    ...    ...   ...     ...  ...    ...   
501  0.06263   0.0  11.93   0.0  0.573  6.593  69.1  2.4786  1.0  273.0   
502  0.04527   0.0  11.93   0.0  0.573  6.120  76.7  2.2875  1.0  273.0   
503  0.06076   0.0  11.93   0.0  0.573  6.976  91.0  2.1675  1.0  273.0   
504  0.10959   0.0  11.93   0.0  0.573  6.794  89.3  2.3889  1.0  273.0   
505  0.04741   0.0  11.93   0.0  0.573  6.030  80.8  2.5050  1.0  273.0   

     PTRATIO       B  LSTAT  
0       15.3  396.90   4.98  
1    

Print all features

In [19]:
print(boston_dataset.feature_names)

['CRIM' 'ZN' 'INDUS' 'CHAS' 'NOX' 'RM' 'AGE' 'DIS' 'RAD' 'TAX' 'PTRATIO'
 'B' 'LSTAT']


In [20]:
df_boston.head(0)

Empty DataFrame
Columns: [CRIM, ZN, INDUS, CHAS, NOX, RM, AGE, DIS, RAD, TAX, PTRATIO, B, LSTAT]
Index: []

In [21]:
print(df_boston.head(0))

Empty DataFrame
Columns: [CRIM, ZN, INDUS, CHAS, NOX, RM, AGE, DIS, RAD, TAX, PTRATIO, B, LSTAT]
Index: []


Print description of the dataset

o/p =

"Boston House Prices dataset\n===========================\n\nNotes\n------\nData Set Characteristics:  \n\n    :Number of Instances: 506 \n\n    :Number of Attributes: 13 numeric/categorical predictive\n    \n    :Median Value (attribute 14) is usually the target\n\n    :Attribute Information (in order):\n        - CRIM     per capita crime rate by town\n        - ZN       proportion of residential land zoned for lots over 25,000 sq.ft.\n        - INDUS    proportion of non-retail business acres per town\n        - CHAS     Charles River dummy variable (= 1 if tract bounds river; 0 otherwise)\n        - NOX      nitric oxides concentration (parts per 10 million)\n        - RM       average number of rooms per dwelling\n        - AGE      proportion of owner-occupied units built prior to 1940\n        - DIS      weighted distances to five Boston employment centres\n        - RAD      index of accessibility to radial highways\n        - TAX      full-value property-tax rate per $10,000\n        - PTRATIO  pupil-teacher ratio by town\n        - B        1000(Bk - 0.63)^2 where Bk is the proportion of blacks by town\n        - LSTAT    % lower status of the population\n        - MEDV     Median value of owner-occupied homes in $1000's\n\n    :Missing Attribute Values: None\n\n    :Creator: Harrison, D. and Rubinfeld, D.L.\n\nThis is a copy of UCI ML housing dataset.\nhttp://archive.ics.uci.edu/ml/datasets/Housing\n\n\nThis dataset was taken from the StatLib library which is maintained at Carnegie Mellon University.\n\nThe Boston house-price data of Harrison, D. and Rubinfeld, D.L. 'Hedonic\nprices and the demand for clean air', J. Environ. Economics & Management,\nvol.5, 81-102, 1978.   Used in Belsley, Kuh & Welsch, 'Regression diagnostics\n...', Wiley, 1980.   N.B. Various transformations are used in the table on\npages 244-261 of the latter.\n\nThe Boston house-price data has been used in many machine learning papers that address regression\nproblems.   \n     \n**References**\n\n   - Belsley, Kuh & Welsch, 'Regression diagnostics: Identifying Influential Data and Sources of Collinearity', Wiley, 1980. 244-261.\n   - Quinlan,R. (1993). Combining Instance-Based and Model-Based Learning. In Proceedings on the Tenth International Conference of Machine Learning, 236-243, University of Massachusetts, Amherst. Morgan Kaufmann.\n   - many more! (see http://archive.ics.uci.edu/ml/datasets/Housing)\n"

In [22]:
boston_dataset.DESCR

".. _boston_dataset:\n\nBoston house prices dataset\n---------------------------\n\n**Data Set Characteristics:**  \n\n    :Number of Instances: 506 \n\n    :Number of Attributes: 13 numeric/categorical predictive. Median Value (attribute 14) is usually the target.\n\n    :Attribute Information (in order):\n        - CRIM     per capita crime rate by town\n        - ZN       proportion of residential land zoned for lots over 25,000 sq.ft.\n        - INDUS    proportion of non-retail business acres per town\n        - CHAS     Charles River dummy variable (= 1 if tract bounds river; 0 otherwise)\n        - NOX      nitric oxides concentration (parts per 10 million)\n        - RM       average number of rooms per dwelling\n        - AGE      proportion of owner-occupied units built prior to 1940\n        - DIS      weighted distances to five Boston employment centres\n        - RAD      index of accessibility to radial highways\n        - TAX      full-value property-tax rate per $10,000

Create a DataFrame from boston.data

In [23]:
df_boston = pandas.DataFrame(data=boston_dataset.data, columns=boston_dataset.feature_names)

In [24]:
df_boston.head(3)

CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD    TAX  \
0  0.00632  18.0   2.31   0.0  0.538  6.575  65.2  4.0900  1.0  296.0   
1  0.02731   0.0   7.07   0.0  0.469  6.421  78.9  4.9671  2.0  242.0   
2  0.02729   0.0   7.07   0.0  0.469  7.185  61.1  4.9671  2.0  242.0   

   PTRATIO       B  LSTAT  
0     15.3  396.90   4.98  
1     17.8  396.90   9.14  
2     17.8  392.83   4.03

Print 5 records from the dataset

In [25]:
df_boston.sample(5)

CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS   RAD    TAX  \
430   8.49213   0.0  18.10   0.0  0.584  6.348  86.1  2.0527  24.0  666.0   
53    0.04981  21.0   5.64   0.0  0.439  5.998  21.4  6.8147   4.0  243.0   
469  13.07510   0.0  18.10   0.0  0.580  5.713  56.7  2.8237  24.0  666.0   
224   0.31533   0.0   6.20   0.0  0.504  8.266  78.3  2.8944   8.0  307.0   
66    0.04379  80.0   3.37   0.0  0.398  5.787  31.1  6.6115   4.0  337.0   

     PTRATIO       B  LSTAT  
430     20.2   83.45  17.64  
53      16.8  396.90   8.43  
469     20.2  396.90  14.76  
224     17.4  385.05   4.14  
66      16.1  396.90  10.24

Add boston.target to your pandas DataFrame with the name PRICE

In [26]:
df_boston["CRIM"].sample(5) 

489    0.18337
54     0.01360
471    4.03841
55     0.01311
93     0.02875
Name: CRIM, dtype: float64

In [27]:
#Here we created a new column into our dataset named : "boston_dataset"
df_boston["PRICE"] = boston_dataset.target 
# price is column name for dF dataFrame
# target is value getting from dataset going into this newly created column("price")

In [28]:
#check its created or not
df_boston.sample(4)

CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS   RAD    TAX  \
396  5.87205   0.0   18.1   0.0  0.693  6.405  96.0  1.6768  24.0  666.0   
234  0.44791   0.0    6.2   1.0  0.507  6.726  66.5  3.6519   8.0  307.0   
54   0.01360  75.0    4.0   0.0  0.410  5.888  47.6  7.3197   3.0  469.0   
311  0.79041   0.0    9.9   0.0  0.544  6.122  52.8  2.6403   4.0  304.0   

     PTRATIO      B  LSTAT  PRICE  
396     20.2  396.9  19.37   12.5  
234     17.4  360.2   8.05   29.0  
54      21.1  396.9  14.80   18.9  
311     18.4  396.9   5.98   22.1

Call the info and describe methods for the dataframe data

o/p is =

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 506 entries, 0 to 505
Data columns (total 14 columns):
CRIM       506 non-null float64
ZN         506 non-null float64
INDUS      506 non-null float64
CHAS       506 non-null float64
NOX        506 non-null float64
RM         506 non-null float64
AGE        506 non-null float64
DIS        506 non-null float64
RAD        506 non-null float64
TAX        506 non-null float64
PTRATIO    506 non-null float64
B          506 non-null float64
LSTAT      506 non-null float64
PRICE      506 non-null float64
dtypes: float64(14)
memory usage: 55.4 KB


In [29]:
df_boston.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 506 entries, 0 to 505
Data columns (total 14 columns):
CRIM       506 non-null float64
ZN         506 non-null float64
INDUS      506 non-null float64
CHAS       506 non-null float64
NOX        506 non-null float64
RM         506 non-null float64
AGE        506 non-null float64
DIS        506 non-null float64
RAD        506 non-null float64
TAX        506 non-null float64
PTRATIO    506 non-null float64
B          506 non-null float64
LSTAT      506 non-null float64
PRICE      506 non-null float64
dtypes: float64(14)
memory usage: 55.5 KB


In [30]:
df_boston.describe

<bound method NDFrame.describe of         CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD    TAX  \
0    0.00632  18.0   2.31   0.0  0.538  6.575  65.2  4.0900  1.0  296.0   
1    0.02731   0.0   7.07   0.0  0.469  6.421  78.9  4.9671  2.0  242.0   
2    0.02729   0.0   7.07   0.0  0.469  7.185  61.1  4.9671  2.0  242.0   
3    0.03237   0.0   2.18   0.0  0.458  6.998  45.8  6.0622  3.0  222.0   
4    0.06905   0.0   2.18   0.0  0.458  7.147  54.2  6.0622  3.0  222.0   
..       ...   ...    ...   ...    ...    ...   ...     ...  ...    ...   
501  0.06263   0.0  11.93   0.0  0.573  6.593  69.1  2.4786  1.0  273.0   
502  0.04527   0.0  11.93   0.0  0.573  6.120  76.7  2.2875  1.0  273.0   
503  0.06076   0.0  11.93   0.0  0.573  6.976  91.0  2.1675  1.0  273.0   
504  0.10959   0.0  11.93   0.0  0.573  6.794  89.3  2.3889  1.0  273.0   
505  0.04741   0.0  11.93   0.0  0.573  6.030  80.8  2.5050  1.0  273.0   

     PTRATIO       B  LSTAT  PRICE  
0       15.3  396.90   4.98 

### Import the XGB Algorithm, mean_squared_error and numpy

bash
conda install py-xgboost

In [31]:
import xgboost
from sklearn.metrics import mean_squared_error
import numpy

In [32]:
df_boston.head(3)

CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD    TAX  \
0  0.00632  18.0   2.31   0.0  0.538  6.575  65.2  4.0900  1.0  296.0   
1  0.02731   0.0   7.07   0.0  0.469  6.421  78.9  4.9671  2.0  242.0   
2  0.02729   0.0   7.07   0.0  0.469  7.185  61.1  4.9671  2.0  242.0   

   PTRATIO       B  LSTAT  PRICE  
0     15.3  396.90   4.98   24.0  
1     17.8  396.90   9.14   21.6  
2     17.8  392.83   4.03   34.7

Select your X and y from the data

In [33]:
x = df_boston.iloc[:,:-1]
#y = df_boston.iloc[0:len(df_boston),-1:len(df_boston)]# this will give me the last only column
y = df_boston.iloc[:,-1]

In [34]:
len(df_boston) #this is for rows 

506

In [35]:
len(df_boston.columns)

14

In [36]:
z = df_boston.iloc[0:len(df_boston),0:-1] #[R:C,R:C] #iloc= integer based location
#y = df_boston.iloc[0:len(df_boston),-1:len(df_boston)]# this will give me the last only column

z.head(3)

CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD    TAX  \
0  0.00632  18.0   2.31   0.0  0.538  6.575  65.2  4.0900  1.0  296.0   
1  0.02731   0.0   7.07   0.0  0.469  6.421  78.9  4.9671  2.0  242.0   
2  0.02729   0.0   7.07   0.0  0.469  7.185  61.1  4.9671  2.0  242.0   

   PTRATIO       B  LSTAT  
0     15.3  396.90   4.98  
1     17.8  396.90   9.14  
2     17.8  392.83   4.03

In [37]:
y.head(3)

0    24.0
1    21.6
2    34.7
Name: PRICE, dtype: float64

### Now you will convert the dataset into an optimized data structure called Dmatrix that XGBoost supports and gives it acclaimed performance and efficiency gains. You will use this later in the tutorial.

In [38]:
# New cmd
data_dmatrix = xgboost.DMatrix(data=x,label=y)

C:\ProgramData\Anaconda3\lib\site-packages\xgboost\core.py:587: FutureWarning: Series.base is deprecated and will be removed in a future version
  if getattr(data, 'base', None) is not None and \
C:\ProgramData\Anaconda3\lib\site-packages\xgboost\core.py:588: FutureWarning: Series.base is deprecated and will be removed in a future version
  data.base is not None and isinstance(data, np.ndarray) \


### Split the data in 30% testing and 70% training 
#common formula since week-3, week-4.pdf

In [39]:
from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(x,y, test_size=0.30, random_state=1029)

### Instantiate an XGBoost regressor object by calling the XGBRegressor() class from the XGBoost library with the hyper-parameters passed as arguments. For classification problems, you would have used the XGBClassifier() class.

In [41]:
myXGBoost = xgboost.XGBRegressor(objective ='reg:linear', colsample_bytree = 0.3, learning_rate = 0.1,
                max_depth = 5, alpha = 10, n_estimators = 10)

### Fit the regressor to the training set and make predictions on the test set using the familiar .fit() and .predict() methods.

o/p = [17:52:42] WARNING: src/objective/regression_obj.cu:152: reg:linear is now deprecated in favor of reg:squarederror.

In [49]:
myXGBoost.fit(x_train, y_train)

[21:39:23] WARNING: src/objective/regression_obj.cu:152: reg:linear is now deprecated in favor of reg:squarederror.


C:\ProgramData\Anaconda3\lib\site-packages\xgboost\core.py:587: FutureWarning: Series.base is deprecated and will be removed in a future version
  if getattr(data, 'base', None) is not None and \


XGBRegressor(alpha=10, base_score=0.5, booster='gbtree', colsample_bylevel=1,
             colsample_bynode=1, colsample_bytree=0.3, gamma=0,
             importance_type='gain', learning_rate=0.1, max_delta_step=0,
             max_depth=5, min_child_weight=1, missing=None, n_estimators=10,
             n_jobs=1, nthread=None, objective='reg:linear', random_state=0,
             reg_alpha=0, reg_lambda=1, scale_pos_weight=1, seed=None,
             silent=None, subsample=1, verbosity=1)

### Print the actual prices from test data

In [52]:
myXGBoost.missing

nan

In [54]:
myXGBoost.learning_rate

0.1

In [56]:
myXGBoost.base_score

0.5

In [64]:
myXGBoost.kwargs

{'alpha': 10}

# Print the actual prices from test data

In [66]:
# Print the actual prices from test data
# x_test data was not used to train the model so this data is pure/untouched data
myXGBoost.predict(x_test)  # so x_test can give us a real time prediction.

array([ 9.613199 , 10.932054 , 12.863521 , 14.545456 , 10.435616 ,
        9.868552 , 10.991049 ,  8.537203 , 10.1254425, 11.781072 ,
       22.830309 , 18.831198 , 13.259726 , 15.512948 , 18.93575  ,
       13.779034 , 16.963024 , 22.611881 , 13.0925255, 16.653694 ,
       15.592541 ,  8.304296 , 11.052871 , 14.394213 , 16.638666 ,
       12.728169 , 19.53332  , 14.984259 , 19.996819 , 17.17081  ,
        9.062304 , 14.937686 , 15.187487 , 18.831198 , 19.154423 ,
       12.384975 , 20.658506 ,  9.787031 ,  7.3919277, 15.046192 ,
       15.583125 , 18.435575 , 18.667007 , 16.061813 , 22.656502 ,
       15.475901 , 13.376813 , 10.999661 , 13.416661 , 14.954578 ,
       10.77917  , 10.749382 , 12.106483 , 11.696469 , 18.952229 ,
       11.25369  , 18.57501  , 16.054462 , 16.173088 ,  7.250405 ,
       12.489716 , 16.35532  ,  6.763396 , 17.010836 , 17.575794 ,
        8.790408 , 22.523949 , 15.4468155, 12.060374 , 23.778217 ,
       22.129562 , 18.18319  , 14.627773 , 13.259726 , 15.3309

# Print the predicted prices

o/p=

array([15. , 26.6, 45.4, 20.8, 34.9, 21.9, 28.7,  7.2, 20. , 32.2, 24.1,
       18.5, 13.5, 27. , 23.1, 18.9, 24.5, 43.1, 19.8, 13.8, 15.6, 50. ,
       37.2, 46. , 50. , 21.2, 14.9, 19.6, 19.4, 18.6, 26.5, 32. , 10.9,
       20. , 21.4, 31. , 25. , 15.4, 13.1, 37.6, 37. , 18.9, 27.9, 50. ,
       14.4, 22. , 19.9, 21.6, 15.6, 15. , 32.4, 29.6, 20.4, 12.3, 19.1,
       14.9, 17.8,  8.8, 35.4, 11.5, 19.6, 20.6, 15.6, 19.9, 23.3, 22.3,
       24.8, 16.1, 22.8, 30.5, 20.4, 24.4, 16.6, 26.2, 16.4, 20.1, 13.9,
       19.4, 22.8, 13.8, 31.6, 10.5, 23.8, 22.4, 19.3, 22.2, 12.6, 19.4,
       22.2, 29.8,  9.6, 34.9, 21.4, 25.3, 32.9, 26.6, 14.6, 31.5, 23.3,
       33.3, 17.5, 19.1, 48.5, 17.1, 23.1, 28.4, 18.9, 13. , 17.2, 24.1,
       18.5, 21.8, 13.3, 23. , 14.1, 23.9, 24. , 17.2, 21.5, 19.1, 20.8,
       36. , 20.1,  8.7, 13.6, 22. , 22.2, 21.1, 13.4, 17.4, 20.1, 10.2,
       23.1, 10.2, 13.1, 14.3, 14.5,  7.2, 19.6, 20.6, 22.7, 26.4,  7.5,
       20.3, 50. ,  8.5, 20.3, 16.1, 22. , 19.6, 10.2, 23.2])

In [76]:
#print(y_test.values.tolist())
y_test.values

array([12.5, 14. , 18.8, 23.7, 10.8,  8.4, 15.6,  9.7, 13.5, 13.1, 50. ,
       23.6, 19. , 22.6, 24.8, 14.4, 22. , 38.7, 14.5, 19.4, 23.3, 16.3,
       14.1, 20.3, 21.9, 19.1, 37. , 29.8, 29. , 24.8, 11.8, 28.7, 19.8,
       33.2, 35.4, 14.8, 33.8, 10.2,  5.6, 21.6, 20.6, 31.5, 31.1, 23.3,
       50. , 25. , 24.5, 17. , 15. , 22.2, 15.4, 19.4, 13.2, 16.8, 32.4,
       15.6, 35.1, 27. , 23.3,  5. , 16.6, 25.2,  8.5, 24.8, 32.5,  8.4,
       50. , 22.9, 15.3, 43.5, 30.1, 28.4, 16.6, 18.5, 23.8, 19.3, 25.3,
       19.1, 27.1, 23.9, 27.5, 32. , 20.4, 13.1, 48.3, 31.5, 16.5, 14.9,
       33.3, 21.7, 21. , 18.4, 33.1, 19. , 23.6, 23.2, 23.1, 34.9, 28.6,
       27.9, 25. , 24.3, 23.4, 14.6, 22.7, 19.8, 36.2, 13.8,  8.7, 50. ,
       11.9, 13.4, 10.2, 23.1, 22. ,  8.8, 15.4, 27.5, 14.1,  7.2, 16.1,
       23. , 19.6, 20.9, 22.1, 24.6, 46. , 20.8, 20. , 25. , 19.5, 28.1,
       33.1,  6.3, 19.4,  7.5, 46.7, 24.1, 21.2, 24.6, 21.7, 18.2, 11.8,
       22. , 34.7, 19.7, 50. , 24.5, 36. , 45.4, 22

# Compare y's

In [79]:
y.head(3)

0    24.0
1    21.6
2    34.7
Name: PRICE, dtype: float64

In [86]:
y = df_boston.iloc[0:len(df_boston),-1:len(df_boston)]
y.head(3)

PRICE
0   24.0
1   21.6
2   34.7

In [88]:
y_test.head(3)

396    12.5
140    14.0
124    18.8
Name: PRICE, dtype: float64

In [103]:
x,y, print("Hi")

Hi


(        CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD    TAX  \
 0    0.00632  18.0   2.31   0.0  0.538  6.575  65.2  4.0900  1.0  296.0   
 1    0.02731   0.0   7.07   0.0  0.469  6.421  78.9  4.9671  2.0  242.0   
 2    0.02729   0.0   7.07   0.0  0.469  7.185  61.1  4.9671  2.0  242.0   
 3    0.03237   0.0   2.18   0.0  0.458  6.998  45.8  6.0622  3.0  222.0   
 4    0.06905   0.0   2.18   0.0  0.458  7.147  54.2  6.0622  3.0  222.0   
 ..       ...   ...    ...   ...    ...    ...   ...     ...  ...    ...   
 501  0.06263   0.0  11.93   0.0  0.573  6.593  69.1  2.4786  1.0  273.0   
 502  0.04527   0.0  11.93   0.0  0.573  6.120  76.7  2.2875  1.0  273.0   
 503  0.06076   0.0  11.93   0.0  0.573  6.976  91.0  2.1675  1.0  273.0   
 504  0.10959   0.0  11.93   0.0  0.573  6.794  89.3  2.3889  1.0  273.0   
 505  0.04741   0.0  11.93   0.0  0.573  6.030  80.8  2.5050  1.0  273.0   
 
      PTRATIO       B  LSTAT  
 0       15.3  396.90   4.98  
 1       17.8  396.90   

In [111]:
x_train, x_test, y_train, y_test

(        CRIM    ZN  INDUS  CHAS    NOX     RM    AGE     DIS   RAD    TAX  \
 470  4.34879   0.0  18.10   0.0  0.580  6.167   84.0  3.0334  24.0  666.0   
 115  0.17134   0.0  10.01   0.0  0.547  5.928   88.2  2.4631   6.0  432.0   
 111  0.10084   0.0  10.01   0.0  0.547  6.715   81.6  2.6775   6.0  432.0   
 114  0.14231   0.0  10.01   0.0  0.547  6.254   84.2  2.2565   6.0  432.0   
 471  4.03841   0.0  18.10   0.0  0.532  6.229   90.7  3.0993  24.0  666.0   
 ..       ...   ...    ...   ...    ...    ...    ...     ...   ...    ...   
 142  3.32105   0.0  19.58   1.0  0.871  5.403  100.0  1.3216   5.0  403.0   
 371  9.23230   0.0  18.10   0.0  0.631  6.216  100.0  1.1691  24.0  666.0   
 192  0.08664  45.0   3.44   0.0  0.437  7.178   26.3  6.4798   5.0  398.0   
 253  0.36894  22.0   5.86   0.0  0.431  8.259    8.4  8.9067   7.0  330.0   
 138  0.24980   0.0  21.89   0.0  0.624  5.857   98.2  1.6686   4.0  437.0   
 
      PTRATIO       B  LSTAT  
 470     20.2  396.90  16.29  


In [112]:
myXGBoost.predict(x_test)

array([ 9.613199 , 10.932054 , 12.863521 , 14.545456 , 10.435616 ,
        9.868552 , 10.991049 ,  8.537203 , 10.1254425, 11.781072 ,
       22.830309 , 18.831198 , 13.259726 , 15.512948 , 18.93575  ,
       13.779034 , 16.963024 , 22.611881 , 13.0925255, 16.653694 ,
       15.592541 ,  8.304296 , 11.052871 , 14.394213 , 16.638666 ,
       12.728169 , 19.53332  , 14.984259 , 19.996819 , 17.17081  ,
        9.062304 , 14.937686 , 15.187487 , 18.831198 , 19.154423 ,
       12.384975 , 20.658506 ,  9.787031 ,  7.3919277, 15.046192 ,
       15.583125 , 18.435575 , 18.667007 , 16.061813 , 22.656502 ,
       15.475901 , 13.376813 , 10.999661 , 13.416661 , 14.954578 ,
       10.77917  , 10.749382 , 12.106483 , 11.696469 , 18.952229 ,
       11.25369  , 18.57501  , 16.054462 , 16.173088 ,  7.250405 ,
       12.489716 , 16.35532  ,  6.763396 , 17.010836 , 17.575794 ,
        8.790408 , 22.523949 , 15.4468155, 12.060374 , 23.778217 ,
       22.129562 , 18.18319  , 14.627773 , 13.259726 , 15.3309

In [152]:
print(x.shape)
print(y.shape)
print(x_test.shape)
print(x_train.shape)
print(y_test.shape)
print(y_train.shape)
print(myXGBoost.predict(x_test).shape)

(506, 13)
(506, 1)
(152, 13)
(354, 13)
(152,)
(354,)
(152,)


### Compute the RMSE by invoking the mean_sqaured_error function from sklearn's metrics module.

o/p = RMSE: 9.715257

In [127]:
mse=numpy.sqrt(mean_squared_error(y_test, myXGBoost.predict(x_test)))
mse

10.40452018557639

In [128]:
import math

In [165]:
rmse= "%f" % mse
rmse

'10.404520'

# To be continue...

https://towardsdatascience.com/linear-regression-on-boston-housing-dataset-f409b7e4a155